### ISSUE TITLE
Slow file processing

### ISSUE THREAD

**Issue by pacija** [AzuraCast/AzuraCast#3500]
Hi,

I rolled out brand new `v0.11 Stable • Ansible` on `Ubuntu 20.04.1 LTS` in VirtualBox VM, with my media library vboxsf-mounted read-only from host into subfolder inside station's media folder (so `.albumart` and `.waveforms` can be writable).

I noticed file processing is really slow. I get around 350 mp3 tracks processed per hour. This means it would take ~5 days to process my collection of 35,000 tracks. I have turned on disk caching in VirtualBox, disk I/O does not seem to be a bottleneck. There's nothing interesting in logs. `top` shows there's plenty of free resources:

```
%Cpu(s):  1.4 us,  0.2 sy,  0.0 ni, 98.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :   1987.6 total,    191.8 free,    345.5 used,   1450.3 buff/cache
```
Is there a way to speed things up a bit? Perhaps this is due to newly introduced creation of waveforms? Perhaps waveform analisys could be split into separate, optional process?

Thank you in advance,

**Comment by NicoxDJ** [comment]
@pacija I opened suggestion for that here: https://azuracast.featureupvote.com/suggestions/102627/option-to-disable-waveforms , but the Admin just ignore my request. :( But you can see what is the workaround what i did to not scan waveforms.

**Comment by pacija** [comment]
@NicoxDJ Thanks! In 48h my fresh AzuraCast instance scanned less than 1/3 of my library, I guess i'll have to nuke it and start anew with your workaround.

I think waveforms are cool feature, but they shouldn't delay the possibility to start broadcasting for so long. I understand waveform processing is demanding task. Even industry leading commercial software like NI Traktor needs almost 24 hours to process my library on fast laptop in parallel processing mode. This consumes so much CPU that Traktor can't be used for reliable live performance while processing.

But processing in NI Traktor is optional, can be selective (process just these tracks), and can be started manually on-demand at the time most convenient for the user, which means I can start using it immediately and process tracks later.

I understand I can't demand for free open source software developed by volunteers to have same features as commercial software, but I think that it would be good design to enable broadcasting as soon as possible, immediately after scraping metadata from tags, while postponing non-essential features like waveform generation. Perhaps waveform generation could be done on-demand, for complete library, or for selected tracks only.

Regards,

**Comment by pacija** [comment]
@NicoxDJ I worked around waveform creation by editing `AudioWaveform.php` as you instructed, with little modification - I put `return [];` at the beginning of `getWaveformFor` function and not at the beginning of whole file, otherwise I was getting errors in `app.log`

```
[2020-12-05T10:11:14.900447+00:00] AzuraCast.ERROR: syntax error, unexpected 'return' (T_RETURN), expecting function (T_FUNCTION) or const (T_CONST) {"file":"/var/azuracast/www/vendor/s
ymfony/messenger/Middleware/HandleMessageMiddleware.php","line":80,"code":0} []
```
Here's how it looks in unified patch:
```
--- /var/azuracast/www/src/Service/AudioWaveform.php.orig      2020-12-05 09:19:32.202691912 +0000
+++ /var/azuracast/www/src/Service/AudioWaveform.php       2020-12-05 13:41:52.868371693 +0000
@@ -15,6 +15,11 @@
      */
     public static function getWaveformFor(string $path): array
     {
+    /**
+     * added by pacija to disable waveforms creation
+     */
+        return [];
+
         if (!file_exists($path)) {
             throw new InvalidArgumentException(sprintf('File at path "%s" not found.', $path));
         }
```
I also noticed that [Command-line Interface](https://www.azuracast.com/developers/cli.html#invoking-the-cli) documentation seems to be incorrect for my setup, and that `cli` command for bare metal installation appears to be `php /var/azuracast/www/bin/console` and not `php /var/azuracast/www/util/cli.php` (there's no such file on that path) or even `php /var/azuracast/www/config/cli.php` (correct path, but does nothing when invoked directly).

I also noticed that `azuracast:media:reprocess` stops at some point, and does not continue when invoked again - it starts from scratch and stops again after some time. Is there a way to make it run until complete library is scanned?

As for media scans, they appear to be triggered by 5-minute tasks, particularly with command:
`php /var/azuracast/www/bin/console queue:process 275`, (as seen in `/var/log/syslog`),which means it will run 25 seconds less than 5 minutes, in order to allow for next scheduled run. It can be invoked manually and passed more than 275 seconds but I don't see any benefit from that.

Regards,

**Comment by Psychoses** [comment]
i have experiment same type of problem since redis change (arround v0.10) on my ansible install. i have found a temp workarround, few days ago, after a lot testing for understand the magic here. if u group cron task seems help. try @ u own risk.

/var/azuracast/www/util/ansible/roles/azuracast-cron/templates/azuracast_cron.j2

https://github.com/AzuraCast/AzuraCast/blob/master/util/ansible/roles/azuracast-cron/templates/azuracast_cron.j2

```
from

#~ 0 * * * * php {{ www_base }}/bin/console sync:run long
#~ */5 * * * * php {{ www_base }}/bin/console sync:run medium
#~ */5 * * * * php {{ www_base }}/bin/console queue:process 275
#~ * * * * * php {{ www_base }}/bin/console sync:run short
#~ * * * * * php {{ www_base }}/bin/console sync:run nowplaying
#~ * * * * * sleep 15; php {{ www_base }}/bin/console sync:run nowplaying
#~ * * * * * sleep 30; php {{ www_base }}/bin/console sync:run nowplaying
#~ * * * * * sleep 45; php {{ www_base }}/bin/console sync:run nowplaying
#~ 0 */6 * * * tmpreaper 12h /var/azuracast/stations/*/temp

to

0 * * * * php {{ www_base }}/bin/console sync:run long
*/5 * * * * php {{ www_base }}/bin/console sync:run medium && php {{ www_base }}/bin/console queue:process 275
* * * * * php {{ www_base }}/bin/console sync:run short
* * * * * php {{ www_base }}/bin/console sync:run nowplaying
* * * * * sleep 15; php {{ www_base }}/bin/console sync:run nowplaying
* * * * * sleep 30; php {{ www_base }}/bin/console sync:run nowplaying
* * * * * sleep 45; php {{ www_base }}/bin/console sync:run nowplaying
0 */6 * * * tmpreaper 12h /var/azuracast/stations/*/temp
```

it's like the "sync:run medium" break the "queue:process 275".

**Comment by SlvrEagle23** [comment]
@pacija @nicoxdj In the interest of optimizing performance, I've switched the default behavior to not pre-generate waveforms for each file during the processing step, instead generating them upon the file's first use in the visual waveform editor (especially since some people don't use this feature at all, it can be a lot of performance deficit for little gain).

I may make this a configurable option in the future, but the one-time short delay of generating waveforms when you first load that page isn't really a huge issue, and given the number of folks who have reported issues with processing time caused by this, I'd rather improve that aspect.

**Comment by NicoxDJ** [comment]
Thanks @SlvrEagle23 ! Going to try it, but the default update.sh leads me to:
```
TASK [azuracast-setup : Run AzuraCast Setup (Update Mode)] **************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "php /var/azuracast/www/bin/console azuracast:setup --update", "delta": "0:00:00.622787", "end": "2020-12-07 15:03:30.336666", "msg": "non-zero return code", "rc": 255, "start": "2020-12-07 15:03:29.713879", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
```
Now i'm doing update.sh --full and will update with the result.

**Comment by SlvrEagle23** [comment]
We have found another issue that was causing much slower file processing; this is fixed as of the latest rolling release version.
